src/lib.rs

```diff
@@ -588,6 +595,182 @@ fn floyd_warshall(py: Python, dag: &digraph::PyDiGraph) -> PyResult<PyObject> {
     Ok(out_dict.into())
 }
 
+fn get_edge_iter_with_weights<G>(
+    graph: G,
+) -> impl Iterator<Item = (usize, usize, PyObject)>
+where
+    G: GraphBase
+        + IntoEdgeReferences
+        + IntoNodeIdentifiers
+        + NodeIndexable
+        + GraphProp
+        + NodesRemoved,
+    G: Data<NodeWeight = PyObject, EdgeWeight = PyObject>,
+{
+    let node_map: Option<HashMap<NodeIndex, usize>>;
+    if graph.nodes_removed() {
+        let mut node_hash_map: HashMap<NodeIndex, usize> = HashMap::new();
+        for (count, node) in graph.node_identifiers().enumerate() {
+            let index = NodeIndex::new(graph.to_index(node));
+            node_hash_map.insert(index, count);
+        }
+        node_map = Some(node_hash_map);
+    } else {
+        node_map = None;
+    }
```

Comment:
In contrast to these lines, why don't we need to care about `node_removed` in `PyGraph::weighted_edge_list()`?

Reply:
You can see the rationale behind why this is needed here: https://github.com/Qiskit/retworkx/pull/80 but basically it comes down to the `StableGraph` type used from petgraph to build the graph data structure in `PyGraph` (and `PyDiGraph`) The `StableGraph` doesn't reuse node indices when there are removals (unless it was the end of the list, which might be a bug actually). So for example, if you had nodes 0, 1, 2, and 3 then removed node 1 you'd have node indices 0, 2, 3. Without mapping 0, 2, 3 to 0, 1, 2 when we build the adjacency matrix we'd either end up with an extra column and row for each removed node or an index error (depending on how we initialized the matrix). The first version of `graph_adjacency_matrix()` had the extra entries in the matrix for missing nodes, but that was decided to be too confusing so this was the compromise to fix it.

For `PyGraph::weighted_edge_list()` this isn't a concern because it just loops over all edges in the graph it doesn't matter if there is a hole in the index list since we'll still get the same list output regardless of whether we're missing nodes or not.

Reply:
I see. Thank you.